src/shortestpaths/johnson.jl

```diff
@@ -0,0 +1,96 @@
+
+"""
+    struct JohnsonState{T, U}
+An [`AbstractPathState`](@ref) designed for Johnson shortest-paths calculations.
+"""
+struct JohnsonState{T<:Real, U<:Integer} <: AbstractPathState
+    dists::Matrix{T}
+    parents::Matrix{U}
+end
+
+
+
+@doc_str """
+johnson_shortest_paths(g, distmx=weights(g); parallel = false)
+Use the [Johnson algorithm](https://en.wikipedia.org/wiki/Johnson%27s_algorithm)
+to compute the shortest paths between all pairs of vertices in graph `g` using an
+optional distance matrix `distmx`.
+If the parameter parallel is set true, dijkstra_shortest_paths will run in parallel.
+Parallel bellman_ford_shortest_paths is currently unavailable
+Return a [`LightGraphs.JohnsonState`](@ref) with relevant
+traversal information.
+Behaviour in case of negative cycle depends on bellman_ford_shortest_paths.
+Throws NegativeCycleError() if a negative cycle is present.
+### Performance
+Complexity: O(|V|*|E|)
+Prefered over Floyd Warshall when graph is sparse.
+If the graph is undirected (and has positive weights) then use Dijkstra.
+Undirected graph with negative weights will throw negative cycle error.
+### Dependencies from LightGraphs
+bellman_ford_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+"""
+function johnson_shortest_paths(
+    g::AbstractGraph{U},
+    distmx::AbstractMatrix{T} = weights(g);
+    parallel::Bool = false
+) where T<:Real where U<:Integer
+
+    nvg = nv(g)
+    #Change when parallel implementation of Bellman Ford available
+    wt_transform = bellman_ford_shortest_paths(g, vertices(g), distmx).dists
+    #Some matrices are immutable
+    new_distmx = convert(SparseMatrixCSC, distmx) 
+
+    #Transform weights
+    for e in edges(g)
+    	new_distmx[src(e), dst(e)] += wt_transform[src(e)] - wt_transform[dst(e)] 
+    end
+    dists = Matrix{T}(nvg, nvg)
+    parents = Matrix{U}(nvg, nvg)
+
+    if !parallel
+    	for v in vertices(g)
+    		dijk_state = dijkstra_shortest_paths(g, v, new_distmx)
+    		dists[v, :] .= dijk_state.dists
+      		parents[v, :] .= dijk_state.parents
+    	end
+    else
+    	dijk_state = parallel_multisource_dijkstra_shortest_paths(g, vertices(g), new_distmx)
+    	dists .= dijk_state.dists
+    	parents .= dijk_state.parents
```

Comment:
see above

Reply:
I'll make the change.
But I was just thinking this step seems wasteful.
It would've been better to copy by reference the data in multi_source_dijkstra_state instead of copy by value.
But I don't think that is allowed for two difference structs with the same field

Reply:
Actually ignore the part about wasteful.
The memory overhead for multiprocess Dijkstra will be much more than the extra matrix.